File: Difficulty: Medium/Kth smallest element/kth-smallest-element.java
//{ Driver Code Starts
import java.io.*;
import java.lang.*;
import java.util.*;

class gfg {
    public static void main(String args[]) {
        // Taking input using class Scanner
        Scanner sc = new Scanner(System.in);

        // Taking count of testcases
        int t = Integer.parseInt(sc.nextLine());
        while (t-- > 0) {
            // taking count of number elements
            String temp[] = sc.nextLine().trim().split(" ");
            int n = temp.length;

            // taking the value of the k
            int k = Integer.parseInt(sc.nextLine());
            // Creating an array of size n
            int arr[] = new int[n];

            // Inserting all the elements to the array
            for (int i = 0; i < n; ++i) arr[i] = Integer.parseInt(temp[i]);

            // calling kthSmallest method of class solve
            // and printing the result
            System.out.println(new Solution().kthSmallest(arr, k));

            System.out.println("~");
        }
    }
}
// } Driver Code Ends


class Solution {
    // Function to find the kth smallest element in the array.
    public static int kthSmallest(int arr[], int k) {
        // Create a priority queue (min-heap) to store elements in ascending order.
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        
        // Add all elements of the array into the priority queue.
        for (int i = 0; i < arr.length; i++) {
            minHeap.add(arr[i]);
        }

        // Remove the smallest element from the heap k-1 times to reach the kth smallest.
        while (k > 1 && !minHeap.isEmpty()) {
            minHeap.poll(); // Removes the smallest element.
            k--; // Decrease k as we've removed one element.
        }

        // The next element is the kth smallest element, so return it.
        return minHeap.poll();
    }
}

File: Difficulty: Medium/Kth smallest element/kth-smallest-element.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    // Function to find the k-th smallest element in the array.
    int kthSmallest(vector<int>& arr, int k) {
        priority_queue<int> maxHeap; // Max heap to store k smallest elements
        
        // Iterate through the array
        for (auto num : arr) {
            maxHeap.push(num); // Add element to the heap

            // Ensure the heap only contains k smallest elements
            if (maxHeap.size() > k) {
                maxHeap.pop();
            }
        }

        return maxHeap.top(); // The k-th smallest element
    }
};


//{ Driver Code Starts.
int main() {
    int t;
    cin >> t; // Number of test cases
    cin.ignore();
    while (t--) {
        // Read the array
        vector<int> arr, arr2;
        string line, line2;
        getline(cin, line);
        stringstream ss(line);
        int num;
        while (ss >> num) {
            arr.push_back(num);
        }
        int k = arr[0];
        getline(cin, line2);
        stringstream ss2(line2);
        while (ss2 >> num) {
            arr2.push_back(num);
        }

        Solution ob;
        cout << ob.kthSmallest(arr2, k) << endl;
        cout << "~\n";
    }
    return 0;
}
// } Driver Code Ends